vindex/cmd/logandmap/main.go

```diff
@@ -0,0 +1,331 @@
+	s, v := getSignerVerifierOrDie()
+
+	// Set up a Tessera POSIX log
+	driver, err := posix.New(ctx, *posixLogDir)
+	if err != nil {
+		return fmt.Errorf("failed to create new log: %v", err)
+	}
+
+	// Get a Tessera appender
+	appender, shutdown, reader, err := tessera.NewAppender(ctx, driver, tessera.NewAppendOptions().
+		WithCheckpointSigner(s).
+		WithBatching(256, time.Second))
+	if err != nil {
+		return fmt.Errorf("failed to get appender: %v", err)
+	}
+	defer func() {
+		_ = shutdown(ctx)
+	}()
+
+	// Create the verifiable index connected to the LogReader.
+	inputLog := logReaderSource{
+		r: reader,
+	}
+	logCpParseFn := func(cpRaw []byte) (*log.Checkpoint, error) {
+		// No witnesses required yet
+		cp, _, _, err := log.ParseCheckpoint(cpRaw, v.Name(), v)
+		return cp, err
+	}
+	vi, err := vindex.NewVerifiableIndex(ctx, inputLog, logCpParseFn, mapFnFromFlags(), walPathFromFlags())
+	if err != nil {
+		return fmt.Errorf("failed to create vindex: %v", err)
+	}
+
+	// Submits new entries to the log in the background.
+	go submitEntries(ctx, appender)
+
+	// Keeps the map synced with the latest published log state.
+	go maintainMap(ctx, vi)
+
+	// Run a web server to handle queries over the verifiable index.
+	go runWebServer(vi)
+	<-ctx.Done()
+	return nil
+}
+
+// logReaderSource adapts a tessera.LogReader to a vindex.InputLog.
+type logReaderSource struct {
+	r tessera.LogReader
+}
+
+func (s logReaderSource) GetCheckpoint(ctx context.Context) (checkpoint []byte, err error) {
+	return s.r.ReadCheckpoint(ctx)
+}
+
+func (s logReaderSource) StreamLeaves(ctx context.Context, start, end uint64) iter.Seq2[[]byte, error] {
+	bi := client.EntryBundles(ctx, 2, s.r.IntegratedSize, s.r.ReadEntryBundle, start, end-start)
+	unbundleFn := func(bundle []byte) ([][]byte, error) {
+		eb := &api.EntryBundle{}
+		if err := eb.UnmarshalText(bundle); err != nil {
+			return nil, err
+		}
+		return eb.Entries, nil
+	}
+
+	return func(yield func([]byte, error) bool) {
+		// Unwrap the client.Entry type to return an iterator of []byte only.
+		for entry, err := range client.Entries(bi, unbundleFn) {
+			if err != nil {
+				if !yield(nil, err) {
+					return
+				}
+				continue
+			}
+			if !yield(entry.Entry, nil) {
+				return
+			}
+		}
+	}
+}
+
+// maintainMap reads entries from the log and sync them to the vindex.
+func maintainMap(ctx context.Context, vi *vindex.VerifiableIndex) {
+	ticker := time.NewTicker(10 * time.Second)
+	defer ticker.Stop()
+
+	for {
+		select {
+		case <-ctx.Done():
+			return
+		case <-ticker.C:
+			if err := vi.Update(ctx); err != nil {
+				klog.Warning(err)
+			}
+		}
+	}
+}
+
+// submitEntries continually creates new log entries and submits them to the log.
+// Entries should be json-encoded LogEntrys.
+// The module should be randomly pulled from a list of [foo, bar, baz, splat]
+// The version should be the current timestamp, as a string
+// The hash should be the sha256 of the module+version
```

Comment:
Have you been vibe-coding? :)

Reply:
That was the goal. But the vibes weren't strong so I ended up doing it myself. Rewritten this to document what it does, rather than what should have been generated.